yt_dlp/postprocessor/sponsorblock.py

```diff
@@ -0,0 +1,391 @@
+        return duration
+
+    def _sanitize_segments(self, duration, segments):
+        segments.sort(key=lambda s: s['segment'][0])
+        # SponsorBlock may deviate from the real one. Sometimes it even reports zero.
+        duration_match = [s for s in segments
+                          if s['videoDuration'] == 0 or abs(duration - s['videoDuration']) <= 1]
+        if len(duration_match) != len(segments):
+            warning = 'Some SponsorBlock segments are from a video of different duration'
+            if not self._force:
+                warning += ', maybe from an old version of this video'
+            else:
+                warning += '. Looks like SponsorBlock has already run'
+            self.report_warning(warning)
+
+        if duration_match:
+            # Ignore milliseconds difference at the start.
+            if duration_match[0]['segment'][0] <= 1:
+                duration_match[0]['segment'][0] = 0
+            # Ignore milliseconds difference at the end.
+            if abs(duration_match[-1]['segment'][1] - duration) <= 1:
+                duration_match[-1]['segment'][1] = duration
+        return duration_match
+
+    def _segments_to_chapters(self, segments):
+        chapters = []
+        # Tracks chapters having the same title to disambiguate them later,
+        # e.g. multiple 'Sponsor' chapters will be renamed to 'Sponsor 1', 'Sponsor 2', etc.
+        unique_titles = {}
+
+        # For overlapping segments, resulting chapter title will look like
+        # '[SponsorBlock]: Sponsor/Interaction Remainder/Self-Promotion', in the order
+        # in which the categories were encountered.
+        # Sets in Python are unordered, and lists take O(N) to check
+        # whether the category was already encountered, thus resorting to OrderedDict.
+        def extract_start_end_cat(s):
+            start, end = s['segment']
+            cat = s['category']
+            return start, end, OrderedDict([(cat, CATEGORY_DISPLAY_NAMES[cat])])
+
+        cur_start, cur_end, cur_cats = extract_start_end_cat(segments[0])
+
+        def append_chapter():
+            title = '[SponsorBlock]: ' + '/'.join(cur_cats.values())
+            chapter = {'start_time': cur_start, 'end_time': cur_end, 'title': title}
+            unique_titles.setdefault(title, []).append(chapter)
+            chapters.append(chapter)
+
+        # Segments returned by SponsorBlock often overlap;
+        # we need to merge segments before converting them into chapters.
+        overlap = False
+        for i in range(1, len(segments)):
+            s = segments[i]
+            start, end = s['segment']
+            if start <= cur_end:
+                overlap = True
+                cur_end, cat = max(cur_end, end), s['category']
+                cur_cats[cat] = CATEGORY_DISPLAY_NAMES[cat]
+            else:
+                append_chapter()
+                cur_start, cur_end, cur_cats = extract_start_end_cat(s)
+        append_chapter()
```

Comment:
If I'm understanding this correctly this'll merge subsegments into one big segment under one category right? The reason I went with the approach I did was so that individual categories could be skipped independently, the user would have less control with your approach if I've understood it correctly.

Reply:
All merged categories are reflected in the title:
```py
title = '[SponsorBlock]: ' + '/'.join(cur_cats.values())
```
Resulting title will look like this: `[SponsorBlock]: Sponsor/Intro/Self-Promotion`.

I give user a warning when overlap is detected:
```py
'Some SponsorBlock segments overlap. '
'Usually it means that there are many conflicting submissions. '
'You may want to review them, and upvote or downvote appropriately'
```

The rationale for this can be seen e.g. here:
(https://sponsor.ajay.app/api/skipSegments?videoID=fFFSKedy9f4&categories=["sponsor","intro","outro","interaction","selfpromo","music_offtopic"]. Someone marked the entire beginning of the video as `sponsor`, which was later refined into separate `intro` and `selfpromo`. The code on its own cannot decide which submissions are correct, that's why the user is warned to review and to upvote/downvote. In this particular case, separate `intro` and `selfpromo` are better, but it may be the other way around for another video.

It is also very hard to make a heuristic: a linear scan through the sorted array will encounter segments in the order `selfpromo`, `sponsor`, and then `intro`, somehow the later `intro` must override the former `sponsor`.

An attempt to create separate entries for each category instead of merging them all together very often results in the creation of milliseconds long chapters. In the linear scan above, `sponsor` ends at 47.68613 while `intro` ends at 48.351925. A separate chapter for `intro` lasts less than a second.

Reply:
Sounds good

>It is also very hard to make a heuristic  
  
Yeah no kidding lol, took me ages to write the cutting code for Sponskrub to deal with this stuff